File: src/pipeline/node/Benchmark.cpp
#include "depthai/pipeline/node/BenchmarkIn.hpp"
#include "depthai/pipeline/node/BenchmarkOut.hpp"

namespace dai {
namespace node {

void BenchmarkOut::setNumMessagesToSend(int num) {
    properties.numMessages = num;
}

void BenchmarkOut::setFps(float fps) {
    properties.fps = fps;
}

void BenchmarkIn::setNumMessagesToGet(int num) {
    properties.numMessages = num;
}

void BenchmarkOut::buildInternal() {
    properties.numMessages = -1;  // By default send messages indefinitely
}

}  // namespace node
}  // namespace dai

File: examples/python/HostNodes/benchmark.py
import depthai as dai
import time

class TestSource(dai.node.ThreadedHostNode):
    def __init__(self, name: str):
        super().__init__()
        self.name = name
        self.output = self.createOutput()

    def run(self):
        while self.isRunning():
            buffer = dai.Buffer()
            print(f"{self.name} node is sending a buffer!")
            self.output.send(buffer)
            time.sleep(1)

with dai.Pipeline(createImplicitDevice=False) as p:
    # Create nodes
    source = TestSource("source")
    benchmarkIn = p.create(dai.node.BenchmarkIn)
    benchmarkIn.setRunOnHost(True)
    benchmarkIn.sendReportEveryNMessages(100)
    benchmarkOut = p.create(dai.node.BenchmarkOut)
    benchmarkOut.setRunOnHost(True)
    benchmarkOut.setFps(30)
    benchmarkOut.out.link(benchmarkIn.input)
    outputQueue = benchmarkIn.report.createOutputQueue()
    source.output.link(benchmarkOut.input)

    p.start()
    while p.isRunning():
        benchmarkReport = outputQueue.get()
        assert isinstance(benchmarkReport, dai.BenchmarkReport)
        print(f"FPS is {benchmarkReport.fps}")